smslibrary/src/main/java/com/eis/smslibrary/SMSMessageParser.java

```diff
@@ -0,0 +1,124 @@
+package com.eis.smslibrary;
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+
+import com.eis.communication.MessageHandler;
+
+import it.lucacrema.preferences.PreferencesManager;
+
+/**
+ * Singleton class used to parse String to SMSMessage and back
+ * Uses a strategy to parse messages, so that any user can update it to its preferred parser
+ * By defaults it uses a default strategy, defined by the library
+ *
+ * @author Luca Crema, Alberto Ursino, Marco Mariotto
+ */
+public class SMSMessageParser implements MessageHandler<String, String, SMSMessage> {
+
+    private static final String PREFERENCE_DEFAULT_STRATEGY = "";
+
+    private SMSParseStrategy parseStrategy;
+    private static SMSMessageParser instance;
+
+    /**
+     * Private constructor
+     * @param context The calling context
+     */
+    private SMSMessageParser(Context context) {
+        SMSParseStrategy savedStrategy = getSavedMessageParseStrategy(context);
+        if(savedStrategy == null) parseStrategy = new DefaultSMSParseStrategy();
+        else parseStrategy = savedStrategy;
+    }
+
+    /**
+     * Method to get the only instance of this class
+     * @param context The calling context
+     * @return {@link SMSMessageParser#instance} the only active instance of this class
+     */
+    public static SMSMessageParser getInstance(Context context){
```

Comment:
Well thought, in this particular case passing a parameter in `getInstance` is better than having a `setup` method because you can only use two other methods that both need the `context`